pkg/fqdn/re/re.go

```diff
@@ -0,0 +1,92 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+// Package re provides a simple function to access compile regex objects for
+// the FQDN subsystem.
+package re
+
+import (
+	"fmt"
+	"regexp"
+	"sync"
+
+	lru "github.com/golang/groupcache/lru"
+
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/option"
+)
+
+// CompileRegex provides a convenience wrapper around the LRU. It will either
+// return the cached regex object or compile the pattern into a regex object if
+// the LRU hasn't been initialized (and initializes it) or it doesn't exist in
+// the cache. If it doesn't exist in the cache, it will be added.
+//
+// This function assumes that it's invoked after the LRU has been initialized
+// via InitRegexCompileLRU(), which happens during the DNS proxy
+// initialization, which assumes that the K8s Watcher is not yet synced and no
+// policies are being processed by the system (meaning no policies to parse,
+// i.e. compiling the FQDN rules or handling FQDN selectors). If that doesn't
+// hold true, then we attempt a best-effort to initialize the LRU.
+func CompileRegex(p string) (*regexp.Regexp, error) {
+	if RegexCompileLRU == nil {
```

Comment:
I think that even the global check here should be protected by the mutex, to ensure that this pointer isn't modified by one goroutine while another goroutine is initializing it?

Reply:
Good point, I've opted to use an `atomic.{Load,Store}Pointer()`, instead of creating a mutex that will only be used once.